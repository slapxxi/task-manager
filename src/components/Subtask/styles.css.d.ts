export const container: string;
export const description: string;
export const input: string;
export const subtasks: string;
export const subtask: string;
export const checkbox: string;
export const completed: string;
export const trashbin: string;
