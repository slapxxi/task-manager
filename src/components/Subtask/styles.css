.container {
  display: flex;
  align-items: center;
}

.description {
  border: 1px solid transparent;
  text-decoration: line-through;
}

.input {
  color: var(--dark-grey);
}

.subtasks {
  margin: 0;
  padding: 0;
  list-style: none;
}

.subtask {
  font-size: 15px;
  padding: 5px 0;
}

.checkbox {
  margin-right: 10px;
}

.completed {
  composes: input;
  color: var(--light-grey);
}

.trashbin {
  --primary-color: hsl(230, 10%, 60%);
}

@media (min-width: 768px) {
  .trashbin {
    visibility: hidden;
  }

  .subtask:hover .trashbin {
    visibility: visible;
  }
}
