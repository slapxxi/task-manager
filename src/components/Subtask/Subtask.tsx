import { Keys } from '@lib';
import { Checkbox, Icon, Input } from '@local/components';
import { Subtask as ISubtask } from '@local/types';
import React, { useEffect, useRef } from 'react';
import trash from '../../assets/trashbin.svg';
import styles from './styles.css';

interface Props {
  subtask: ISubtask;
  focus?: boolean;
  onEdit?: (subtask: ISubtask) => void;
  onFocus?: (subtask: ISubtask) => void;
  onBlur?: (subtask: ISubtask) => void;
  onSubmit?: (subtask: ISubtask) => void;
  onRemove?: (subtask: ISubtask) => void;
}

function Subtask({ subtask, focus, onEdit, onFocus, onBlur, onRemove, onSubmit }: Props) {
  const inputRef = useRef<HTMLInputElement>();

  useEffect(
    () => {
      if (focus) {
        inputRef.current.focus();
      }
    },
    [focus],
  );

  function handleToggle(value: boolean) {
    if (onEdit) {
      onEdit({ ...subtask, completed: value });
    }
  }

  function handleFocus() {
    if (onFocus) {
      onFocus(subtask);
    }
  }

  function handleBlur() {
    if (onBlur) {
      onBlur(subtask);
    }
  }

  function handleEdit(e: React.ChangeEvent<HTMLInputElement>) {
    if (onEdit) {
      onEdit({ ...subtask, description: e.currentTarget.value });
    }
  }

  function handleRemove() {
    if (onRemove) {
      onRemove(subtask);
    }
  }

  function handleKeyDown(e: React.KeyboardEvent<HTMLInputElement>) {
    const keyCode = e.keyCode || e.charCode;
    if (e.currentTarget.value === '' && keyCode === Keys.backspace) {
      handleRemove();
      return;
    }
    if (keyCode === Keys.enter && onSubmit) {
      onSubmit(subtask);
      return;
    }
  }

  return (
    <div className={styles.container}>
      <Checkbox
        value={subtask.completed}
        size={15}
        onToggle={handleToggle}
        className={styles.checkbox}
      />
      <Input
        type="text"
        className={subtask.completed && styles.completed}
        value={subtask.description}
        placeholder="What should be done..."
        onFocus={handleFocus}
        onBlur={handleBlur}
        onChange={handleEdit}
        onKeyDown={handleKeyDown}
        ref={inputRef}
        data-testid="input"
      />
      <Icon glyph={trash} size={16} onClick={handleRemove} className={styles.trashbin} />
    </div>
  );
}

export default React.memo(Subtask);
