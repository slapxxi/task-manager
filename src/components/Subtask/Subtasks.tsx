import { Subtask as ISubtask } from '@local/types';
import * as React from 'react';
import styles from './styles.css';

interface Props {
  subtasks: ISubtask[];
  render?: (params: { subtask: ISubtask; index: number }) => React.ReactNode;
}

function Subtasks({ subtasks, render }: Props) {
  return (
    <ul className={styles.subtasks}>
      {subtasks.map((st, index) => (
        <li key={st.id} className={styles.subtask}>
          {render && render({ subtask: st, index })}
        </li>
      ))}
    </ul>
  );
}

export default React.memo(Subtasks);
