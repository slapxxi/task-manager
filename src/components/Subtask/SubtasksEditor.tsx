import { createSubtask } from '@lib';
import { Subtask as ISubtask } from '@local/types';
import isEmpty from 'lodash-es/isEmpty';
import React, { useRef } from 'react';
import Subtask from './Subtask';
import Subtasks from './Subtasks';

interface Props {
  subtasks: ISubtask[];
  focus?: boolean;
  onCreate?: (subtask: ISubtask) => void;
  onEdit?: (subtask: ISubtask) => void;
  onRemove?: (subtask: ISubtask) => void;
}

function SubtasksEditor({ subtasks, focus, onCreate, onEdit, onRemove }: Props) {
  const lastFocusedRef = useRef<number>(NaN);

  function handleEdit(subtask: ISubtask) {
    if (onEdit) {
      onEdit(subtask);
    }
  }

  function handleCreate(subtask: ISubtask) {
    if (onCreate) {
      onCreate(createSubtask(subtask));
    }
  }

  function handleAdd() {
    if (onCreate) {
      onCreate(createSubtask({}));
      lastFocusedRef.current = subtasks.length;
    }
  }

  function handleRemove(subtask: ISubtask) {
    if (onRemove) {
      if (subtasks.length === 1) {
        lastFocusedRef.current = 0;
      } else {
        lastFocusedRef.current -= 1;
      }
      onRemove(subtask);
    }
  }

  function handleFocus(index: number) {
    lastFocusedRef.current = index;
  }

  function handleBlur() {
    lastFocusedRef.current = NaN;
  }

  return (
    <div>
      <Subtasks
        subtasks={isEmpty(subtasks) ? [createSubtask({})] : subtasks}
        render={({ subtask, index }) =>
          isEmpty(subtasks) && index === subtasks.length ? (
            <Subtask subtask={subtask} onEdit={handleCreate} focus />
          ) : (
            <Subtask
              focus={focus && lastFocusedRef.current === index}
              subtask={subtask}
              onEdit={handleEdit}
              onFocus={() => handleFocus(index)}
              onBlur={handleBlur}
              onRemove={handleRemove}
              onSubmit={handleAdd}
            />
          )
        }
      />
    </div>
  );
}

export default SubtasksEditor;
