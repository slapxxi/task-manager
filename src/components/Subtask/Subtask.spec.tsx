import React from 'react';
import { fireEvent, render } from 'react-testing-library';
import Subtask from './Subtask';

jest.mock('@local/components/IconSystem/IconSystem');

const subtask = { id: 'test', description: 'test', completed: false };

it('renders', () => {
  const { container } = render(<Subtask subtask={subtask} />);
  expect(container.firstChild).toMatchSnapshot();
});

it('invokes callback when editing', () => {
  const spy = jest.fn();
  const { getByTestId } = render(<Subtask subtask={subtask} onEdit={spy} />);
  fireEvent.change(getByTestId('input'), { target: { value: 'new' } });
  expect(spy).toHaveBeenCalledWith({
    id: 'test',
    description: 'new',
    completed: false,
  });
});

it('invokes callback when subtask removed', () => {
  const spy = jest.fn();
  const { getByTestId } = render(<Subtask subtask={subtask} onRemove={spy} />);
  fireEvent.click(getByTestId('remove-test'));
  expect(spy).toHaveBeenCalledTimes(1);
  expect(spy).toHaveBeenCalledWith({
    id: 'test',
    description: 'test',
    completed: false,
  });
});
