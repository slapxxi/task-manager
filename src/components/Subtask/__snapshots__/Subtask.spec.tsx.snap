// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div
  class="container"
>
  <div
    class="checkbox"
  >
    <div
      style="position: relative; display: flex; align-items: center; height: 15px;"
    >
      <input
        class="input"
        type="checkbox"
      />
      <svg
        clip-rule="evenodd"
        height="15"
        preserveAspectRatio="none"
        viewBox="0 0 100 100"
        width="15"
      >
        <circle
          cx="50"
          cy="50"
          fill="none"
          r="44"
          stroke="#99a"
          stroke-width="8"
        />
      </svg>
    </div>
  </div>
  <input
    class="input sc-bwzfXH PGmuR"
    data-testid="input"
    placeholder="What should be done..."
    type="text"
    value="test"
  />
  <button
    class="sc-bdVaJa cabFfC"
    data-testid="remove-test"
  >
    <div
      class="trashbin"
      name="trashbin"
      size="16"
    >
      icon-
      trashbin
    </div>
  </button>
</div>
`;
